pkg/server/authorizedentries/cache_test.go

```diff
@@ -0,0 +1,441 @@
+			test.assertAuthorizedEntries(t, agent1)
+		})
+
+		t.Run("agent has selector match", func(t *testing.T) {
+			// Workload entry is available through alias since the agent
+			// has a non-strict superset of the alias selectors.
+			test.assertAuthorizedEntries(t, agent2, workloadEntry)
+		})
+
+		t.Run("agent has strict selector superset", func(t *testing.T) {
+			// Workload entry is available through alias since the agent
+			// has a strict superset of the alias selectors.
+			test.assertAuthorizedEntries(t, agent3, workloadEntry)
+		})
+	})
+
+	t.Run("alias removed", func(t *testing.T) {
+		var (
+			aliasEntry    = makeAlias(alias1, sel1, sel2)
+			workloadEntry = makeWorkload(alias1)
+		)
+
+		cache := testCache().
+			withEntries(workloadEntry, aliasEntry).
+			withAgent(agent1, sel1, sel2).
+			hydrate()
+
+		cache.RemoveEntry(aliasEntry.Id)
+		assertAuthorizedEntries(t, cache, agent1)
+	})
+
+	t.Run("agent removed", func(t *testing.T) {
+		var (
+			aliasEntry    = makeAlias(alias1, sel1, sel2)
+			workloadEntry = makeWorkload(alias1)
+		)
+
+		cache := testCache().
+			withEntries(workloadEntry, aliasEntry).
+			withAgent(agent1, sel1, sel2).
+			hydrate()
+
+		cache.RemoveAgent(agent1.String())
+		assertAuthorizedEntries(t, cache, agent1)
+	})
+
+	t.Run("agent pruned after expiry", func(t *testing.T) {
+		var (
+			aliasEntry    = makeAlias(alias1, sel1, sel2)
+			workloadEntry = makeWorkload(alias1)
+		)
+
+		cache := testCache().
+			withEntries(workloadEntry, aliasEntry).
+			withExpiredAgent(agent1, sel1, sel2).
+			hydrate()
+		assertAuthorizedEntries(t, cache, agent1, workloadEntry)
+
+		assert.Equal(t, 1, cache.PruneExpiredAgents())
+
+		assertAuthorizedEntries(t, cache, agent1)
+	})
+}
+
+func TestCacheInternalStats(t *testing.T) {
+	// This test asserts that the internal indexes are properly maintained
+	// across various operations. The motivation is to ensure that as the cache
+	// is updated that we are appropriately inserting and removing records from
+	// the indexees.
+	t.Run("pristine", func(t *testing.T) {
+		cache := NewCache()
+		require.Zero(t, cache.stats())
+	})
+
+	t.Run("entries and aliases", func(t *testing.T) {
+		entry1 := makeWorkload(agent1)
+		entry2a := makeWorkload(agent2)
+
+		// Version b will change to an alias instead
+		entry2b := makeAlias(alias1, sel1, sel2)
+		entry2b.Id = entry2a.Id
+
+		cache := NewCache()
+		cache.UpdateEntry(entry1)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  1,
+			EntriesByParentID: 1,
+		}, cache.stats())
+
+		cache.UpdateEntry(entry2a)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  2,
+			EntriesByParentID: 2,
+		}, cache.stats())
+
+		cache.UpdateEntry(entry2b)
+		require.Equal(t, cacheStats{
+			EntriesByEntryID:  1,
+			EntriesByParentID: 1,
+			AliasesByEntryID:  2, // one for each selector
+			AliasesBySelector: 2, // one for each selector
+		}, cache.stats())
+
+		cache.RemoveEntry(entry1.Id)
```

Comment:
may we add a test case that when removing an invalid entryID nothing change?

Reply:
Sure, done.